compiler/optimizer/VPHandlers.cpp

```diff
@@ -4758,7 +4758,7 @@ TR::Node * constrainLongNumberOfLeadingZeros(OMR::ValuePropagation *vp, TR::Node
    return constrainHighestOneBitAndLeadingZerosHelper (vp, node, getLongConst,
                                                        getLongRange, getLong, getLowHighLongs,
                                                        createIntConstConstraint,
-                                                       createIntRangeConstraint, longNumberOfLeadingZeros, (int64_t) 0, (int64_t) -1);
+                                                       createLongRangeConstraint, longNumberOfLeadingZeros, (int64_t) 0, (int64_t) -1);
```

Comment:
This to me looks like an actual bug, so pointing it out here. Here is the warning:

```
[2021-05-28T22:32:43.924Z]     32>C:\jenkins\workspace\PullRequest-win_x86-64\Build\compiler\optimizer\VPHandlers.cpp(4604): warning C4244: 'argument' : conversion from 'int64_t' to 'int32_t', possible loss of data [C:\jenkins\workspace\PullRequest-win_x86-64\Build\build\jitbuilder\jitbuilder.vcxproj]
[2021-05-28T22:32:43.924Z]                  C:\jenkins\workspace\PullRequest-win_x86-64\Build\compiler\optimizer\VPHandlers.cpp(4678) : see reference to function template instantiation 'void addValidRangeBlockOrGlobalConstraint<TR::VPConstraint*(__cdecl *)(OMR::ValuePropagation *,int32_t,int32_t),int32_t(__cdecl *)(int64_t),T>(OMR::ValuePropagation *,TR::Node *,FUNC,FUNC2,T,T,bool)' being compiled
[2021-05-28T22:32:43.924Z]                  with
[2021-05-28T22:32:43.924Z]                  [
[2021-05-28T22:32:43.924Z]                      T=int64_t,
[2021-05-28T22:32:43.924Z]                      FUNC=TR::VPConstraint *(__cdecl *)(OMR::ValuePropagation *,int32_t,int32_t),
[2021-05-28T22:32:43.924Z]                      FUNC2=int32_t (__cdecl *)(int64_t)
[2021-05-28T22:32:43.924Z]                  ]
[2021-05-28T22:32:43.924Z]                  C:\jenkins\workspace\PullRequest-win_x86-64\Build\compiler\optimizer\VPHandlers.cpp(4761) : see reference to function template instantiation 'TR::Node *constrainHighestOneBitAndLeadingZerosHelper<TR::VPLongConst*(__cdecl *)(TR::VPConstraint *),TR::VPLongRange*(__cdecl *)(TR::VPConstraint *),void(__cdecl *)(TR::VPLongConst *,int64_t &),void(__cdecl *)(TR::VPLongRange *,int64_t &,int64_t &),TR::VPConstraint*(__cdecl *)(OMR::ValuePropagation *,int32_t),TR::VPConstraint*(__cdecl *)(OMR::ValuePropagation *,int32_t,int32_t),int32_t(__cdecl *)(int64_t),int64_t>(OMR::ValuePropagation *,TR::Node *,A,B,C,D,E,F,G,T,T)' being compiled
[2021-05-28T22:32:43.924Z]                  with
[2021-05-28T22:32:43.924Z]                  [
[2021-05-28T22:32:43.924Z]                      A=TR::VPLongConst *(__cdecl *)(TR::VPConstraint *),
[2021-05-28T22:32:43.924Z]                      B=TR::VPLongRange *(__cdecl *)(TR::VPConstraint *),
[2021-05-28T22:32:43.924Z]                      C=void (__cdecl *)(TR::VPLongConst *,int64_t &),
[2021-05-28T22:32:43.924Z]                      D=void (__cdecl *)(TR::VPLongRange *,int64_t &,int64_t &),
[2021-05-28T22:32:43.924Z]                      E=TR::VPConstraint *(__cdecl *)(OMR::ValuePropagation *,int32_t),
[2021-05-28T22:32:43.924Z]                      F=TR::VPConstraint *(__cdecl *)(OMR::ValuePropagation *,int32_t,int32_t),
[2021-05-28T22:32:43.924Z]                      G=int32_t (__cdecl *)(int64_t),
[2021-05-28T22:32:43.924Z]                      T=int64_t
[2021-05-28T22:32:43.924Z]                  ]
```

It seems like we're trying to create a constraint with integer range, but the types here are all longs.

Reply:
`lnolz` is generated only in one place that I can see, which is to replace `invokestatic Long.numberOfLeadingZeros(J)I` [in OpenJ9's IL generator](https://github.com/eclipse-openj9/openj9/blob/0c8f6b1861f99b63b99824b0befcc5a4869c1f40/runtime/compiler/ilgen/Walker.cpp#L3718-L3719). That method returns a 32-bit integer rather than a 64-bit one, which is perfectly sensible considering that the maximum possible result is only 64. And the [IL opcode properties confirm that the result is `Int32`](https://github.com/eclipse/omr/blob/eeb1035617c5b0beea82d81bfda9eed89e8da743/compiler/il/OMROpcodes.enum#L9261-L9262) rather than `Int64` (though I have no idea why they don't specify the type of the child...)

Since the result is 32-bit, we really should create "int" constraints for it, so I think the warning needs to be fixed some other way

Reply:
Thanks for calling attention to this one btw!

Reply:
Fixed in b2e3cb5. Did it in a separate commit to flag it out and separate the change given the size of the PR.